plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/XBuildManager.java

```diff
@@ -247,27 +250,28 @@ private XBuildable doGetIncrementalBuildable(List<URI> dirtyFiles, List<URI> del
 		try {
 			Map<ProjectDescription, Set<URI>> project2dirty = computeProjectToUriMap(this.dirtyFiles);
 			Map<ProjectDescription, Set<URI>> project2deleted = computeProjectToUriMap(this.deletedFiles);
+			SetView<ProjectDescription> changedPDs = Sets.union(project2dirty.keySet(), project2deleted.keySet());
+			IOrderInfo<ProjectDescription> orderInfo = projectBuildOrderProvider.get(changedPDs);
 
-			SetView<ProjectDescription> allDescriptions = Sets.union(project2dirty.keySet(), project2deleted.keySet());
-			List<ProjectDescription> sortedDescriptions = sortByDependencies(allDescriptions);
-
-			for (ProjectDescription descr : sortedDescriptions) {
+			for (ProjectDescription descr : orderInfo) {
```

Comment:
There is quite some logic behind the iterator that is implicitly created here, esp. considering the fact that this iterator will be influenced by the `#visitAffected()` call inside the for loop's block and that several iterators may be created from a single iterable. Please try to simplify this and make it a bit more explicit.